python/az/aro/azext_aro/_aad.py

```diff
@@ -71,3 +78,35 @@ def create_service_principal(self, app_id):
                 retries += 1
                 logger.warning("%s; retry %d of %d", ex, retries, max_retries)
                 time.sleep(10)
+
+    def refresh_application_credentials(self, object_id):
+        password = uuid.uuid4()
+        key_id = uuid.uuid4()
+        start_date = datetime.datetime.utcnow()
+        end_date = datetime.datetime(2299, 12, 31, tzinfo=datetime.timezone.utc)
+
+        try:
+            credentials = list(self.client.applications.list_password_credentials(object_id))
+        except GraphErrorException as e:
+            logger.error(e.message)
+            raise
+
+        # when appending credentials ALL fields must be present, otherwise
+        # azure gives ambiguous errors about not being able to update old keys
+        credentials.append(PasswordCredential(
+            custom_key_identifier=str(start_date).encode(),  # bytearray
+            key_id=str(key_id),
+            start_date=start_date,
+            end_date=end_date,
+            value=password))
+
+        # keys created with older version of cli are not updatable
+        # https://github.com/Azure/azure-sdk-for-python/issues/18131
+        try:
+            self.client.applications.update_password_credentials(object_id, credentials)
+        except GraphErrorException:
+            raise BadRequestError("failed to update an application.\n \
+Please consider using the command below to trigger credential rotation manually: \n \
+az aro update --name MyCluster --resource-group MyResourceGroup --client-id ClientId --client-secret ClientSecret")
```

Comment:
how about something like this:
* specifically look for the bad case where the custom_key_identifier is empty
* if that happens, raise a BadRequestError with the following text:

Cluster AAD application contains a client secret with an empty description.  Please either manually remove the existing client secret and run `az aro update --refresh-cluster-service-principal`, or manually create a new client secret and run `az aro update --client-secret`.

* if not, and you get a GraphErrorException, just raise it.